File: heron/simulator/src/java/com/twitter/heron/simulator/grouping/ReduceGrouping.java
//  Copyright 2017 Twitter. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License
package com.twitter.heron.simulator.grouping;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.twitter.heron.proto.system.HeronTuples;

public class ReduceGrouping extends Grouping {
  private static Logger LOG = Logger.getLogger(ReduceGrouping.class.getName());
  private final int taskIdsSize;
  private int nextTaskIndex = 0;

  public ReduceGrouping(List<Integer> taskIds) {
    super(taskIds);

    taskIdsSize = taskIds.size();
    nextTaskIndex = new Random().nextInt(taskIds.size());
  }

  @Override
  public List<Integer> getListToSend(HeronTuples.HeronDataTuple tuple) {
    List<Integer> res = new ArrayList<>();
    if (tuple.getSubTaskDest()) {
//      LOG.log(Level.INFO, String.format("ReduceGrouping %d -> %d",
//          tuple.getSourceTask(), tuple.getDestTaskIdsList().get(0)));
      res.add(tuple.getDestTaskIdsList().get(0));
    } else {
//      LOG.log(Level.INFO, String.format("Reduce Random Grouping %d -> %d",
//          tuple.getSourceTask(), taskIds.get(nextTaskIndex)));
      res.add(taskIds.get(nextTaskIndex));
      nextTaskIndex = getNextTaskIndex(nextTaskIndex);
    }
    return res;
  }

  private int getNextTaskIndex(int index) {
    int dupIndex = index + 1;

    if (dupIndex == taskIdsSize) {
      dupIndex = 0;
    }

    return dupIndex;
  }
}

File: heron/stmgr/src/cpp/grouping/reduce-grouping.cpp
/*
 * Copyright 2015 Twitter, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "grouping/reduce-grouping.h"
#include <functional>
#include <iostream>
#include <list>
#include <vector>
#include "grouping/grouping.h"
#include "proto/messages.h"
#include "basics/basics.h"
#include "errors/errors.h"
#include "threads/threads.h"
#include "network/network.h"

namespace heron {
namespace stmgr {

ReduceGrouping::ReduceGrouping(const std::vector<sp_int32>& _task_ids) : Grouping(_task_ids) {
  next_index_ = rand() % task_ids_.size();
}

ReduceGrouping::~ReduceGrouping() {}

void ReduceGrouping::GetListToSend(const proto::system::HeronDataTuple& _tuple,
                                    std::vector<sp_int32>& _return) {
  if (_tuple.sub_task_dest() && _tuple.dest_task_ids_size() > 0) {
    // LOG(INFO) << "Reduction to: " << _tuple.dest_task_ids(0);
    _return.push_back(_tuple.dest_task_ids(0));
  } else {
    _return.push_back(task_ids_[next_index_]);
    next_index_ = (next_index_ + 1) % task_ids_.size();
//    LOG(INFO) << "shuffle to: " << next_index_;
  }
}

}  // namespace stmgr
}  // namespace heron
